### PR TITLE
Support local files for ReadDir operations

### PR DESCRIPTION
**Pull request by ashmeenkaur**
### Description
Added support to list local files present in a directory.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - 
 * ls command on mounted directory with local files
 * ls command on explicit dir with local files
 * ls command on implicit dir with local files
 * ls -R command

here 
`l1, l2, l3, l4, l5` and `l6` are local files.
`c` and `d` are explicit directories
`foo` is implicit directory
error `ls: cannot access 'l*': No such file or directory` is because changes are required in lookupInode command.

 ```
 ashmeen@ashmeen:~/Documents/mnt$ ls -R
.:
ls: cannot access 'l1': No such file or directory
ls: cannot access 'l2': No such file or directory
c  d  foo  l1  l2

./c:
ls: cannot access './c/l3': No such file or directory
l3

./d:
ls: cannot access './d/l4': No such file or directory
ls: cannot access './d/l5': No such file or directory
l4  l5

./foo:
ls: cannot access './foo/l6': No such file or directory
bar  l6

 ```
3. Unit tests - Added
4. Integration tests - NA

### CODE REVIEW

File: internal/fs/dir_handle.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ package fs
 
 import (
 	"fmt"
+	"path"
```
Comment (vadlakondaswetha) [inline internal/fs/dir_handle.go:19]: can you try moving dir_handle.go into handle package and write unit tests.
Having dir_handle in fs package doesn't make sense.

File: internal/fs/dir_handle.go [hunk 2]
```diff
@@ -159,8 +160,10 @@ func fixConflictingNames(entries []fuseutil.Dirent) (err error) {
 //
 // LOCKS_REQUIRED(in)
 func readAllEntries(
-	ctx context.Context,
-	in inode.DirInode) (entries []fuseutil.Dirent, err error) {
+		ctx context.Context,
+		in inode.DirInode,
+		localFileInodes map[inode.Name]inode.Inode) (entries []fuseutil.Dirent, err error) {
+	//Read entries from GCS
```
Comment (vadlakondaswetha) [inline internal/fs/dir_handle.go:166]: nit: For all comments add space at the begining and full stop at the end.

File: internal/fs/local_file_test.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	f1, err := os.Create(path.Join(mntDir, "localFile1"))
+	AssertEq(nil, err)
+	f2, err := os.Create(path.Join(mntDir, "localFile2"))
+	AssertEq(nil, err)
+
+	// Attempt to list mntDir.
+	entries, err := os.ReadDir(mntDir)
+	AssertEq(nil, err)
+	info, err := entries[0].Info()
+	fmt.Println(entries[0].IsDir(), entries[0].Name(), entries[0].Type(), info)
```
Comment (vadlakondaswetha) [inline internal/fs/local_file_test.go:65]: nit: remove fmt.Println statements when you raise PR

File: internal/fs/local_file_test.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+////////////////////////////////////////////////////////////////////////
+// Tests
+////////////////////////////////////////////////////////////////////////
+
+func (t *LocalFileTest) TestReadDirWithEmptyLocalFile() {
+	// Create local files.
+	f1, err := os.Create(path.Join(mntDir, "localFile1"))
+	AssertEq(nil, err)
+	f2, err := os.Create(path.Join(mntDir, "localFile2"))
```
Comment (vadlakondaswetha) [inline internal/fs/local_file_test.go:58]: Please validate that file isnot created in GCS and created when file.close is called.

reference: https://github.com/GoogleCloudPlatform/gcsfuse/pull/1279/files

File: internal/fs/local_file_test.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+				// File
+				"implicitFoo/bar": "",
+			}))
+	f1, err := os.Create(path.Join(mntDir, "implicitFoo/file1"))
+	AssertEq(nil, err)
+	f2, err := os.Create(path.Join(mntDir, "implicitFoo/file2"))
+	AssertEq(nil, err)
+
+	// Attempt to list explicit directory.
+	entries, err := os.ReadDir(path.Join(mntDir, "implicitFoo/"))
```
Comment (vadlakondaswetha) [inline internal/fs/local_file_test.go:128]: do you want to extract all this common logic into seperate method.

**Review by vadlakondaswetha — CHANGES_REQUESTED** [review]
Please fix all lint errors

File: internal/fs/dir_handle.go [hunk 6]
```diff
@@ -218,15 +229,30 @@ func readAllEntries(
 	return
 }
 
+// localFileEntries lists the local files present in the directory.
+// Local means that the file is not yet present on GCS.
+func localFileEntries(d inode.DirInode, localFileInodes map[inode.Name]inode.Inode) (localEntries []fuseutil.Dirent, err error) {
```
Comment (vadlakondaswetha) [inline internal/fs/dir_handle.go:234]: Thinking loud,  when we implement that rmDir and renameDir doesnt support when localfiles are present, we might need this logic in fs.go. See if you need this logic and if yes move this to fs.go

Comment (ashmeenkaur) [reply internal/fs/dir_handle.go:234]: I initially added this method to DirInode, but realised that all the methods in DirInode today deals with GCS objects. That will be the most appropriate location from code reusability POV as well. Do you think it will be okay to move this to DirInode?

Comment (vadlakondaswetha) [reply internal/fs/dir_handle.go]: Yeah should be good, if that operations have dirInode with them.

File: internal/fs/handle/dir_handle_test.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+		&t.clock,
+		true) // localFile
+	return
+}
+
+////////////////////////////////////////////////////////////////////////
+// Tests
+////////////////////////////////////////////////////////////////////////
+
+func (t *DirHandleTest) EnsureEntriesWithLocalAndGCSFiles() {
```
Comment (vadlakondaswetha) [inline internal/fs/handle/dir_handle_test.go:108]: Test cases when 
1. local fileinode and GCS object has same name
2. local fileinode and GCS directory has same name

File: internal/fs/inode/dir.go [hunk 8]
```diff
@@ -126,6 +126,10 @@ type DirInode interface {
 		ctx context.Context,
 		name string,
 		isImplicitDir bool) (err error)
+
+	// localFileEntries lists the local files present in the directory.
+	// Local means that the file is not yet present on GCS.
+	LocalFileEntries(localFileInodes map[Name]Inode) (localEntries []fuseutil.Dirent)
```
Comment (vadlakondaswetha) [inline internal/fs/inode/dir.go:132]: should we call it getLocalFileEntries or ListLocalFileEntries. what is the go recommendation?

File: internal/fs/inode/dir_test.go [hunk 9]
```diff
@@ -1210,3 +1229,30 @@ func (t *DirTest) CreateLocalChildFile_ShouldnotCreateObjectInGCS() {
...
+
+func (t *DirTest) LocalFileEntriesEmpty() {
+	localFileInodes := map[Name]Inode{}
+
+	entries := t.in.LocalFileEntries(localFileInodes)
+
+	AssertEq(0, len(entries))
+}
+
+func (t *DirTest) LocalFileEntriesWith2LocalChildFiles() {
```
Comment (vadlakondaswetha) [inline internal/fs/inode/dir_test.go:1241]: Add unit test when all non child files are passed. this is to ensure that empty list is returned and not error or nil

File: internal/fs/local_file_test.go [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+		WriteConfig: config.WriteConfig{CreateEmptyFile: false},
+	}
+	t.fsTest.SetUpTestSuite()
+}
+
+////////////////////////////////////////////////////////////////////////
+// Tests
+////////////////////////////////////////////////////////////////////////
+
+func (t *LocalFileTest) TestReadDirWithEmptyLocalFile() {
```
Comment (vadlakondaswetha) [inline internal/fs/local_file_test.go:56]: can you add test case for listing non-empty files and checkinng if the size is returned correctly.
This is our main usecase. The purpose of supporting ls is to show the size of the file.

File: internal/fs/local_file_test.go [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+	// Verify entries received successfully.
+	AssertEq(2, len(entries))
+	t.verifyLocalFileEntry(entries[0], fileName1)
+	t.verifyLocalFileEntry(entries[1], fileName2)
+	// Close the local files.
+	t.closeFileAndValidateObjectContents(f1, fileName1, "")
+	t.closeFileAndValidateObjectContents(f2, fileName2, "")
+}
+
+func (t *LocalFileTest) TestReadDirForExplicitDirWithLocalFile() {
```
Comment (vadlakondaswetha) [inline internal/fs/local_file_test.go:75]: is it possible to write test for ls -R?

**Review by vadlakondaswetha — CHANGES_REQUESTED** [review]

File: internal/fs/inode/dir.go (hunk 8 above)
Comment (ashmeenkaur) [reply internal/fs/inode/dir.go:132]: The recommendation is to avoid "Get" prefix and to prefer starting the name with the noun directly go/go-style/decisions#getters

File: internal/fs/local_file_test.go (hunk 10 above)
Comment (ashmeenkaur) [reply internal/fs/local_file_test.go:56]: This test case requires lookUpInode to be supported for local files. I'll add this test in a separate PR.

Comment (ashmeenkaur) [reply internal/fs/local_file_test.go:96]: added a TODO for the same. [link](https://github.com/GoogleCloudPlatform/gcsfuse/pull/1287/files#diff-7800585ad28fc6c239f5a8d71b053eab502f7f40d5c0c7d6b794d4a4c3386afcR211)

File: internal/fs/local_file_test.go [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+}
+
+////////////////////////////////////////////////////////////////////////
+// Helpers
+////////////////////////////////////////////////////////////////////////
+
+func (t *LocalFileTest) createLocalFile(filePath string) (file *os.File) {
+	file, err := os.Create(filePath)
+	AssertEq(nil, err)
+	t.validateObjectNotFoundErr(filePath)
```
[the remark this answers is not in the document]
Comment (ashmeenkaur) [reply internal/fs/local_file_test.go:61]: path to mounted directory (mntDir) is not required in the path for GCS objects. It is only required for local files here.

I noticed that i was using path.join in every call to createLocalFile, so moving that logic to this method.

[the remark this answers is not in the document]
Comment (vadlakondaswetha) [reply internal/fs/local_file_test.go]: I didnt follow this. I am saying
os.Create should have /mntDir also in the path
and t.validateObjectNotFoundErr(filePath) shouldn't have /mntDir in the path.

Right now you are using same parameter fileName in both methods.

File: internal/fs/handle/dir_handle_test.go [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+	t.validateEntry(t.dh.entries[0], "gcsObject1", fuseutil.DT_File)
+	t.validateEntry(t.dh.entries[1], "gcsObject2", fuseutil.DT_File)
+	t.validateEntry(t.dh.entries[2], "localFile1", fuseutil.DT_File)
+	t.validateEntry(t.dh.entries[3], "localFile2", fuseutil.DT_File)
+}
+
+func (t *DirHandleTest) EnsureEntriesWithSameNameLocalAndGCSFile() {
+	var err error
+	// Set up empty GCS objects.
+	_, err = gcsutil.CreateObject(t.ctx, t.bucket, "mnt/file1", nil)
```
[the remark this answers is not in the document]
Comment (ashmeenkaur) [reply internal/fs/handle/dir_handle_test.go:187]: Mount is not the mounted directory. I should have used a better name here. Updating to use "directory" here instead.

[the remark this answers is not in the document]
Comment (ashmeenkaur) [reply internal/fs/handle/dir_handle_test.go]: "directory" is a directory present on gcs. 
DirHandle holds a DirInode pointing to this "directory"

File: internal/fs/local_file_test.go (hunk 10 above)
Comment (ashmeenkaur) [reply internal/fs/local_file_test.go:248]: Added the test

File: internal/fs/local_file_test.go [hunk 14]
```diff
@@ -177,29 +245,135 @@ func (t *LocalFileTest) RandomWritesToLocalFile() {
...
-	AssertEq(nil, err)
-	t.validateObjectNotFoundErr(fileName)
+	// Recursively list mntDir/subDir/ directory.
+	cmd := exec.Command("ls", "-R", path.Join(mntDir, "subDir/"))
+	out, err := cmd.Output()
 
-	return
+	// Verify output.
+	AssertEq(nil, err)
+	AssertTrue(strings.Contains(string(out), "/subDir:\nbaseLocalFile\nexplicitFoo\nimplicitFoo"))
```
Comment (vadlakondaswetha) [inline internal/fs/local_file_test.go:373]: can you try using filepath.walk or filepath.walkDir to recursively all files under directory.  Creating one cache for all buckets
Refactoring the code to use same file-cache implementation

We can avoid assertion on a string

File: internal/fs/local_file_test.go [hunk 15]
```diff
@@ -177,29 +255,177 @@ func (t *LocalFileTest) RandomWritesToLocalFile() {
...
+	t.closeFileAndValidateObjectContents(f1, "implicitFoo/"+FileName, "")
+	t.closeFileAndValidateObjectContents(f2, "implicitFoo/"+FileName2, "")
+}
 
-	return
+func (t *LocalFileTest) TestRecursiveListingWithLocalFiles() {
+	/* Structure
+	mntDir/
+		- baseLocalFile 			--- file
+		- explicitFoo/ 				--- directory
```
Comment (vadlakondaswetha) [inline internal/fs/local_file_test.go:343]: implicit??

File: internal/fs/local_file_test.go [hunk 16]
```diff
@@ -177,29 +255,177 @@ func (t *LocalFileTest) RandomWritesToLocalFile() {
...
+			return err
+		}
+		// The object type is not directory.
+		if !dir.IsDir() {
+			return nil
+		}
+
+		objs, err := os.ReadDir(path)
+		if err != nil {
+			log.Fatal(err)
```
Comment (vadlakondaswetha) [inline internal/fs/local_file_test.go:382]: why only this log.Fatal?

File: internal/fs/local_file_test.go [hunk 17]
```diff
@@ -177,29 +255,177 @@ func (t *LocalFileTest) RandomWritesToLocalFile() {
...
+
+		// Check if mntDir/implicitFoo/ has correct objects.
+		if path == mntDir+"/implicitFoo" {
+			// numberOfObjects = 2
+			if len(objs) != 2 {
+				return fmt.Errorf("incorrect number of objects in mntDir/implicitFoo/")
+			}
+
+			if objs[0].Name() != "bar" || objs[0].IsDir() != false ||
+				objs[1].Name() != implicitLocalFileName || objs[1].IsDir() != false {
```
Comment (vadlakondaswetha) [inline internal/fs/local_file_test.go:419]: can we use Assert statements here similar to your other statements?

File: internal/fs/local_file_test.go (hunk 16 above)
Comment (ashmeenkaur) [reply internal/fs/local_file_test.go]: Changed to assert instead

File: internal/fs/local_file_test.go (hunk 17 above)
Comment (ashmeenkaur) [reply internal/fs/local_file_test.go]: done

**Review by vadlakondaswetha — APPROVED** [review]

**Review by vadlakondaswetha — DISMISSED** [review]

File: internal/fs/handle/dir_handle.go [hunk 18]
```diff
@@ -182,11 +184,26 @@ func readAllEntries(
...
 	// below.
 	sort.Sort(sortedDirents(entries))
 
 	// Fix name conflicts.
+	// When a local file is synced to GCS but not removed from the local file map,
+	// the entries list will have two duplicate entries.
+	// To handle this scenario, we had 2 options:
+	// Option 1: [Selected]
+	// Throw an error while fixing conflicting names. The error will be fixed in
+	// subsequent ls calls.
```
Comment (vadlakondaswetha) [inline internal/fs/handle/dir_handle.go:201]: this error will be fixed in subsequent ls calls assuming that entry will be removed from localFileInodes.

**Review by vadlakondaswetha — DISMISSED** [review]

**Review by vadlakondaswetha — APPROVED** [review]
